modules/ws/src/main/java/org/mule/module/ws/consumer/MuleWSDLLocator.java

```diff
@@ -0,0 +1,163 @@
+import org.mule.api.MuleException;
+import org.mule.module.http.api.requester.proxy.ProxyConfig;
+import org.mule.transport.http.HttpConnector;
+import org.mule.transport.ssl.api.TlsContextFactory;
+import org.mule.util.IOUtils;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.Collection;
+
+import javax.wsdl.WSDLException;
+import javax.wsdl.xml.WSDLLocator;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.xml.sax.InputSource;
+
+/**
+ * A custom locator to replicate the context of requester in every step 
+ * where wsdl4j has to retrieve a resource (imported XSD, WSDL)
+ */
+public class MuleWSDLLocator implements WSDLLocator
+{
+    private static final Logger logger = LoggerFactory.getLogger(WSConsumer.class);
+
+    private String baseURI;
+    private String latestImportedURI;
+    private boolean useConnectorToRetrieveWsdl;
+    private HttpRequesterWsdlRetrieverStrategy httpRetrieverStrategy;
+    private URLRetrieverStrategy urlRetrieverStrategy;
+    private Collection<InputStream> streams = new ArrayList<InputStream>();
+    
+    public MuleWSDLLocator(MuleWSDLLocatorConfig config) throws MuleException
+    {
+        this.baseURI = config.getBaseURI();
+        this.useConnectorToRetrieveWsdl = config.isUseConnectorToRetrieveWsdl();
+        createURLRetrieverStrategy();
+        if (useConnectorToRetrieveWsdl)
+        {
+            createHttpRequesterWsdlRetrieverStrategy(config.getTlsContextFactory(), config.getProxyConfig(), config.getContext());
+        }
+    }
+
+
+    private void createURLRetrieverStrategy()
+    {
+        urlRetrieverStrategy = new URLRetrieverStrategy();
+
+    }
+
+
+    private void createHttpRequesterWsdlRetrieverStrategy(TlsContextFactory tlsContextFactory, ProxyConfig proxyConfig, MuleContext context) throws MuleException
+    {
+        httpRetrieverStrategy = new HttpRequesterWsdlRetrieverStrategy(tlsContextFactory, proxyConfig, context);
+        httpRetrieverStrategy.startHttpClient();
+    }
+
+
+    @Override
+    public InputSource getBaseInputSource()
+    {
+        try
+        {
+            return getInputSource(baseURI);
+        }
+        catch (Exception e)
+        {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public InputSource getImportInputSource(String parentLocation, String importLocation)
+    {
+
+        try
+        {
+            latestImportedURI = IOUtils.getResourceAsUrl(importLocation, getClass()).toString();
+
+            return getInputSource(importLocation);
+        }
+        catch (Exception e)
+        {
+            throw new RuntimeException(e);
+        }
+
+    }
+
+    private InputSource getInputSource(String url) throws WSDLException
+    {
+        boolean isHttpRequester = verifyIsHttpRequester(url);
+
+        InputStream resultStream = null;
+
+        if (useConnectorToRetrieveWsdl && isHttpRequester)
```

Comment:
Now it is arranged through an abstract pattern. I get a concrete factory according to the strategy to use, create the strategy and then I use the interface method to retrieveWsdl.